File: LC007ReverseInteger.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/reverse-integer/
 */
public class LC007ReverseInteger {
    public class Solution {
        public int reverse(int x) {
            long big = x, result = 0;
            int sign = 1;
            if (x < 0) {
                big = -big;
                sign = -1;
            }
            while (big > 0) {
                result *= 10;
                result += big % 10;
                big /= 10;
            }
            result *= sign;
            if (result < Integer.MIN_VALUE || Integer.MAX_VALUE < result) {
                return 0;
            }
            return (int) result;
        }
    }

    @Test
    public void test_1000000003() throws Exception {
        assertEquals(0, new Solution().reverse(1000000003));
    }

    @Test
    public void test_1534236469() throws Exception {
        assertEquals(0, new Solution().reverse(1534236469));
    }

    @Test
    public void test_Minus321() throws Exception {
        assertEquals(-123, new Solution().reverse(-321));
    }
}

File: algorithms/leetcode/lc007_reverse_integer/LC007ReverseInteger.cs
// https://leetcode.com/problems/reverse-integer/

using System;

namespace AlgorithmDesign.leetcode.lc007_reverse_integer
{
    public class Solution
    {
        public int Reverse(int x)
        {
            if (x == Int32.MinValue)
            {
                return 0;
            }
            bool negative = x < 0;
            if (negative)
            {
                x = -x;
            }
            int[] digits = new int[10];
            int collected = 0;
            while (x > 0)
            {
                digits[collected++] = x % 10;
                x /= 10;
            }
            int[] limits = new int[] { 2, 1, 4, 7, 4, 8, 3, 6, 4, 7 };
            int length = limits.Length;
            int padding = length - collected;
            Array.Copy(digits, 0, digits, padding, collected);
            Array.Fill(digits, 0, 0, padding);
            for (int i = 0; i < length; i++)
            {
                if (digits[i] > limits[i])
                {
                    return 0;
                }
                if (digits[i] < limits[i])
                {
                    break;
                }
            }
            int result = 0;
            foreach (int digit in digits)
            {
                result = result * 10 + digit;
            }
            if (negative)
            {
                result = -result;
            }
            return result;
        }
    }
}
